### PR TITLE
Box3: Refactored .setFromObject()

### PR DESCRIPTION
**Pull request by Mugen87**
see #10296

Implements the suggestion of @WestLangley

### CODE REVIEW

**Comment by WestLangley** [comment]
This looks good.

You know, we have
```javascript
Vector3.fromAttribute( attribute, index, offset )
```
It would be nice to be able to use it -- or something similar -- instead of calling `getX()`, `getY()`, `getZ()`.

**Comment by Mugen87** [comment]
Oh, `fromAttribute` is new to me. But does it actually work in this context? `InterleavedBufferAttribute` has a different implementation of `getX()`, `getY()` and `getZ()` than `BufferAttribute`. 

If you just call `getX()` in `Box3. setFromObject ()`, it's not necessary to know, which kind of `BufferGeometry` is used. `Vector3.fromAttribute` means you have to explore the type of buffer so you can correctly setup the parameters, right?

**Comment by Mugen87** [comment]
`Vector3.fromAttribute` does this:
```javascript
this.x = attribute.array[ index ];
```
This does not work with `InterleavedBufferAttribute` because the array is nested in the `data` property. Getting values is done like this:
```javascript
attribute.data.array[ index * attribute.data.stride + attribute.offset ];
```

**Comment by Mugen87** [comment]
But i don't think it's good to reveal these implementation details of `InterleavedBufferAttribute` in `Vector3`...

**Comment by Mugen87** [comment]
Maybe we could to this...

```javascript
fromAttribute: function ( attribute, index ) {

	this.x = attribute.getX( index );
	this.y = attribute.getY( index );
	this.z = attribute.getZ( index );

	return this;

}
```

**Comment by WestLangley** [comment]
Hmm... I don't remember why I added the `offset` arg...
```javascript
Vector3.fromAttribute( attribute, index, offset )
```

**Comment by Mugen87** [comment]
`.fromAttribute()` in `Vector2`, `Vector3` and `Vector4` now works with `InterleavedBufferAttribute`. So the method is safe to use in `Box3 .setFromObject()`

@WestLangley Do you think the changes in `Vector*` are okay like this?

File: src/math/Box3.js [hunk 1]
```diff
@@ -165,9 +165,7 @@ Box3.prototype = {
 
 							for ( i = 0, l = attribute.count; i < l; i ++ ) {
 
-								v1.x = attribute.getX( i );
-								v1.y = attribute.getY( i );
-								v1.z = attribute.getZ( i );
+								v1.fromAttribute( attribute, i );
```
Comment (WestLangley) [inline src/math/Box3.js]: Chaining is OK. (here and elsewhere...)
```javascript
v1.fromAttribute( attribute, i ).applyMatrix4( node.matrixWorld );
```

Comment (Mugen87) [inline src/math/Box3.js]: 👍

File: src/math/Vector2.js [hunk 2]
```diff
@@ -453,14 +453,10 @@ Vector2.prototype = {
 
 	},
 
-	fromAttribute: function ( attribute, index, offset ) {
-
-		if ( offset === undefined ) offset = 0;
-
-		index = index * attribute.itemSize + offset;
+	fromAttribute: function ( attribute, index ) {
```
Comment (WestLangley) [inline src/math/Vector2.js]: Warn `if ( offset !== undefined )`. Here and elsewhere.

**Comment by Mugen87** [comment]
I hope the warning text is okay...

**Comment by WestLangley** [comment]
I'm glad you are revisiting this...

Questions for another PR: 

Should `Box3.setFromBufferAttribute()` be using `expandByPoint()`?

Compare `Box3.setFromArray()` vs. `Box3.setFromPoints()`.

See https://github.com/mrdoob/three.js/pull/7974 regarding performance issues.

**Comment by Mugen87** [comment]
@mrdoob Do you think the changes are okay like this? It seems the new code of `Vector .fromAttribute()` is currently the only pragmatic way for supporting `InterleavedBufferAttribute`...

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**